In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import * 

In [ ]:
import numpy as np

In [ ]:
import pandas as pd

In [ ]:
data_csvfile = pd.read_csv('lab6_ex1.csv')
print(data_csvfile)

  Warehouses/Markets  Hassan  Jodhpur  Trichy  Simla  Vijayawada  Dibrugarh  \
0          Ahmedabad     827      817    1470   2272        1615       3143   
1          Bengaluru     559     1832     751   3772         637       3773   
2            Chennai     110     2443     577   3865         798       3521   
3              Delhi    2943     1471    4676   1337        2702       2741   
4            Kolkata    3742     3089    2931   3199        2226       1739   
5            Lucknow    2353     1160    3295    368        2369       2365   
6             Mumbai    1356     1617    2045   2941        2334       4151   
7          Hyderabad     195     1862    1331   2921         402       2825   
8             Demand     205      300     425    270         375        225   

   Raipur  Agartala  Available Stock  
0    1264      4465              350  
1    1856      5459              185  
2    1984      5303              305  
3    2482      4148              180  
4    1691     

In [ ]:
data_csvfile.columns

Index(['Warehouses/Markets', 'Hassan', 'Jodhpur', 'Trichy', 'Simla',
       'Vijayawada', 'Dibrugarh', 'Raipur', 'Agartala', 'Available Stock'],
      dtype='object')

In [ ]:
# create a model
model_lab6_ex1 = ConcreteModel()

In [ ]:
row_indices = range(M)
col_indices = range(N)

print(row_indices)
print(col_indices)

range(0, 8)
range(0, 8)


In [ ]:
#to get the contents of costs alone from the file we use the following
for row_idx in row_indices:
  for col_idx in col_indices:
    print('cost[',row_idx,'][',col_idx+1,']:',data_csvfile.iloc[row_idx, col_idx+1])

cost[ 0 ][ 1 ]: 827
cost[ 0 ][ 2 ]: 817
cost[ 0 ][ 3 ]: 1470
cost[ 0 ][ 4 ]: 2272
cost[ 0 ][ 5 ]: 1615
cost[ 0 ][ 6 ]: 3143
cost[ 0 ][ 7 ]: 1264
cost[ 0 ][ 8 ]: 4465
cost[ 1 ][ 1 ]: 559
cost[ 1 ][ 2 ]: 1832
cost[ 1 ][ 3 ]: 751
cost[ 1 ][ 4 ]: 3772
cost[ 1 ][ 5 ]: 637
cost[ 1 ][ 6 ]: 3773
cost[ 1 ][ 7 ]: 1856
cost[ 1 ][ 8 ]: 5459
cost[ 2 ][ 1 ]: 110
cost[ 2 ][ 2 ]: 2443
cost[ 2 ][ 3 ]: 577
cost[ 2 ][ 4 ]: 3865
cost[ 2 ][ 5 ]: 798
cost[ 2 ][ 6 ]: 3521
cost[ 2 ][ 7 ]: 1984
cost[ 2 ][ 8 ]: 5303
cost[ 3 ][ 1 ]: 2943
cost[ 3 ][ 2 ]: 1471
cost[ 3 ][ 3 ]: 4676
cost[ 3 ][ 4 ]: 1337
cost[ 3 ][ 5 ]: 2702
cost[ 3 ][ 6 ]: 2741
cost[ 3 ][ 7 ]: 2482
cost[ 3 ][ 8 ]: 4148
cost[ 4 ][ 1 ]: 3742
cost[ 4 ][ 2 ]: 3089
cost[ 4 ][ 3 ]: 2931
cost[ 4 ][ 4 ]: 3199
cost[ 4 ][ 5 ]: 2226
cost[ 4 ][ 6 ]: 1739
cost[ 4 ][ 7 ]: 1691
cost[ 4 ][ 8 ]: 2128
cost[ 5 ][ 1 ]: 2353
cost[ 5 ][ 2 ]: 1160
cost[ 5 ][ 3 ]: 3295
cost[ 5 ][ 4 ]: 368
cost[ 5 ][ 5 ]: 2369
cost[ 5 ][ 6 ]: 2365
cost[ 5 ][ 7 ]: 1293
cost[ 5 ][ 8 ]: 3642
c

In [ ]:
#declare the decision variables in the model
model_lab6_ex1.x = Var(row_indices, col_indices, domain= NonNegativeIntegers)

In [ ]:
model_lab6_ex1.objective = Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab6_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

In [ ]:
model_lab6_ex1.constraints = ConstraintList()

In [ ]:
#first we will add the available stock constraints
for i in range(M):
  model_lab6_ex1.constraints.add(expr=sum(model_lab6_ex1.x[i,j] for j in col_indices) <= data_csvfile.iloc[i,N+1])
  
#Now we will add the demand constraints
for j in range(N):
  model_lab6_ex1.constraints.add(expr=sum(model_lab6_ex1.x[i,j] for i in row_indices) == data_csvfile.iloc[M,j+1])

In [ ]:
model_lab6_ex1.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   16 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
    x_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :   64 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (6, 0), (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6), (6, 7), (7, 0), (7, 1), (7, 2), (7, 3), (7, 4), (7, 5), (7, 6), (7, 7)}
    x_index_0 : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
    

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model_lab6_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
/# display solution
print('\nObjective = ', model_lab6_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Number of truckloads of EVs between',data_csvfile['Warehouses/Markets'][i],'and', data_csvfile.columns[j+1],':', model_lab6_ex1.x[i,j].value)

print('\nConstraints')
model_lab6_ex1.constraints.display()


Objective =  2192870.0

Decision Variables
Number of truckloads of EVs between Ahmedabad and Hassan : 0.0
Number of truckloads of EVs between Ahmedabad and Jodhpur : 300.0
Number of truckloads of EVs between Ahmedabad and Trichy : 0.0
Number of truckloads of EVs between Ahmedabad and Simla : 0.0
Number of truckloads of EVs between Ahmedabad and Vijayawada : 0.0
Number of truckloads of EVs between Ahmedabad and Dibrugarh : 45.0
Number of truckloads of EVs between Ahmedabad and Raipur : 5.0
Number of truckloads of EVs between Ahmedabad and Agartala : 0.0
Number of truckloads of EVs between Bengaluru and Hassan : 0.0
Number of truckloads of EVs between Bengaluru and Jodhpur : 0.0
Number of truckloads of EVs between Bengaluru and Trichy : 120.0
Number of truckloads of EVs between Bengaluru and Simla : 0.0
Number of truckloads of EVs between Bengaluru and Vijayawada : 65.0
Number of truckloads of EVs between Bengaluru and Dibrugarh : 0.0
Number of truckloads of EVs between Bengaluru and Ra

In [ ]:
data_csvfile.iloc[5,4]=10000000000

In [ ]:
model_lab6_ex1.objective = Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab6_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
result = opt_cbc.solve(model_lab6_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
/# display solution
print('\nObjective = ', model_lab6_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Number of truckloads of EVs between',data_csvfile['Warehouses/Markets'][i],'and', data_csvfile.columns[j+1],':', model_lab6_ex1.x[i,j].value)

print('\nConstraints')
model_lab6_ex1.constraints.display()


Objective =  2380570.0

Decision Variables
Number of truckloads of EVs between Ahmedabad and Hassan : 0.0
Number of truckloads of EVs between Ahmedabad and Jodhpur : 300.0
Number of truckloads of EVs between Ahmedabad and Trichy : 0.0
Number of truckloads of EVs between Ahmedabad and Simla : 0.0
Number of truckloads of EVs between Ahmedabad and Vijayawada : 0.0
Number of truckloads of EVs between Ahmedabad and Dibrugarh : 0.0
Number of truckloads of EVs between Ahmedabad and Raipur : 50.0
Number of truckloads of EVs between Ahmedabad and Agartala : 0.0
Number of truckloads of EVs between Bengaluru and Hassan : 0.0
Number of truckloads of EVs between Bengaluru and Jodhpur : 0.0
Number of truckloads of EVs between Bengaluru and Trichy : 120.0
Number of truckloads of EVs between Bengaluru and Simla : 0.0
Number of truckloads of EVs between Bengaluru and Vijayawada : 65.0
Number of truckloads of EVs between Bengaluru and Dibrugarh : 0.0
Number of truckloads of EVs between Bengaluru and Ra

In [ ]:
data_csvfile.iloc[4,6]=10000000000

In [ ]:
data_csvfile.iloc[3,7]=10000000000

In [ ]:
model_lab6_ex1.objective = Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab6_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
result = opt_cbc.solve(model_lab6_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
/# display solution
print('\nObjective = ', model_lab6_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Number of truckloads of EVs between',data_csvfile['Warehouses/Markets'][i],'and', data_csvfile.columns[j+1],':', model_lab6_ex1.x[i,j].value)

print('\nConstraints')
model_lab6_ex1.constraints.display()


Objective =  2518200.0

Decision Variables
Number of truckloads of EVs between Ahmedabad and Hassan : 0.0
Number of truckloads of EVs between Ahmedabad and Jodhpur : 300.0
Number of truckloads of EVs between Ahmedabad and Trichy : 0.0
Number of truckloads of EVs between Ahmedabad and Simla : 0.0
Number of truckloads of EVs between Ahmedabad and Vijayawada : 0.0
Number of truckloads of EVs between Ahmedabad and Dibrugarh : 50.0
Number of truckloads of EVs between Ahmedabad and Raipur : 0.0
Number of truckloads of EVs between Ahmedabad and Agartala : 0.0
Number of truckloads of EVs between Bengaluru and Hassan : 0.0
Number of truckloads of EVs between Bengaluru and Jodhpur : 0.0
Number of truckloads of EVs between Bengaluru and Trichy : 120.0
Number of truckloads of EVs between Bengaluru and Simla : 0.0
Number of truckloads of EVs between Bengaluru and Vijayawada : 65.0
Number of truckloads of EVs between Bengaluru and Dibrugarh : 0.0
Number of truckloads of EVs between Bengaluru and Ra

**Answer 1:**

Minimize 827$x_{0,0}$ + 817$x_{0,1}$ + 1470$x_{0,2}$ + 2272$x_{0,3}$ + 1615$x_{0,4}$ + 3143$x_{0,5}$ + 1264$x_{0,6}$ + 4465$x_{0,7}$ + 559$x_{1,0}$ + 1832$x_{1,1}$ + 751$x_{1,2}$ + 3772$x_{1,3}$ + 637$x_{1,4}$ + 3773$x_{1,5}$ + 1856$x_{1,6}$ + 5459$x_{1,7}$ + 110$x_{2,0}$ + 2443$x_{2,1}$ + 577$x_{2,2}$ + 3865$x_{2,3}$ + 798$x_{2,4}$ + 3521$x_{2,5}$ + 1984$x_{2,6}$ + 5303$x_{2,7}$ + 2943$x_{3,0}$ + 1471$x_{3,1}$ + 4676$x_{3,2}$ + 1337$x_{3,3}$ + 2702$x_{3,4}$ + 2741$x_{3,5}$ + 2482$x_{3,6}$ + 4148$x_{3,7}$ + 3742$x_{4,0}$ + 3089$x_{4,1}$ + 2931$x_{4,2}$ + 3199$x_{4,3}$ + 2226$x_{4,4}$ + 1739$x_{4,5}$ + 1691$x_{4,6}$ + 2128$x_{4,7}$ + 2353$x_{5,0}$ + 1160$x_{5,1}$ + 3295$x_{5,2}$ + 368$x_{5,3}$ + 2369$x_{5,4}$ + 2365$x_{5,5}$ + 1293$x_{5,6}$ + 3642$x_{5,7}$ + 1356$x_{6,0}$ + 1617$x_{6,1}$ + 2045$x_{6,2}$ + 2941$x_{6,3}$ + 2334$x_{6,4}$ + 4151$x_{6,5}$ + 1937$x_{6,6}$ + 4287$x_{6,7}$ + 195$x_{7,0}$ + 1862$x_{7,1}$ + 1331$x_{7,2}$ + 2921$x_{7,3}$ + 402$x_{7,4}$ + 2825$x_{7,5}$ + 563$x_{7,6}$ + 4441$x_{7,7}$

subject to constraints 

$x_{0,0}$ + $x_{0,1}$ + $x_{0,2}$ + $x_{0,3}$ + $x_{0,4}$ + $x_{0,5}$ + $x_{0,6}$ + $x_{0,7}$ $\leq$ 350.0

$x_{1,0}$ + $x_{1,1}$ + $x_{1,2}$ + $x_{1,3}$ + $x_{1,4}$ + $x_{1,5}$ + $x_{1,6}$ + $x_{1,7}$ $\leq$ 185.0

$x_{2,0}$ + $x_{2,1}$ + $x_{2,2}$ + $x_{2,3}$ + $x_{2,4}$ + $x_{2,5}$ + $x_{2,6}$ + $x_{2,7}$ $\leq$ 305.0

$x_{3,0}$ + $x_{3,1}$ + $x_{3,2}$ + $x_{3,3}$ + $x_{3,4}$ + $x_{3,5}$ + $x_{3,6}$ + $x_{3,7}$ $\leq$ 180.0

$x_{4,0}$ + $x_{4,1}$ + $x_{4,2}$ + $x_{4,3}$ + $x_{4,4}$ + $x_{4,5}$ + $x_{4,6}$ + $x_{4,7}$ $\leq$ 205.0

$x_{5,0}$ + $x_{5,1}$ + $x_{5,2}$ + $x_{5,3}$ + $x_{5,4}$ + $x_{5,5}$ + $x_{5,6}$ + $x_{5,7}$ $\leq$ 175.0

$x_{6,0}$ + $x_{6,1}$ + $x_{6,2}$ + $x_{6,3}$ + $x_{6,4}$ + $x_{6,5}$ + $x_{6,6}$ + $x_{6,7}$ $\leq$ 325.0

$x_{7,0}$ + $x_{7,1}$ + $x_{7,2}$ + $x_{7,3}$ + $x_{7,4}$ + $x_{7,5}$ + $x_{7,6}$ + $x_{7,7}$ $\leq$ 370.0

$x_{0,0}$ + $x_{1,0}$ + $x_{2,0}$ + $x_{3,0}$ + $x_{4,0}$ + $x_{5,0}$ + $x_{6,0}$ + $x_{7,0}$ = 205.0

$x_{0,1}$ + $x_{1,1}$ + $x_{2,1}$ + $x_{3,1}$ + $x_{4,1}$ + $x_{5,1}$ + $x_{6,1}$ + $x_{7,1}$ = 300.0

$x_{0,2}$ + $x_{1,2}$ + $x_{2,2}$ + $x_{3,2}$ + $x_{4,2}$ + $x_{5,2}$ + $x_{6,2}$ + $x_{7,2}$ = 425.0

$x_{0,3}$ + $x_{1,3}$ + $x_{2,3}$ + $x_{3,3}$ + $x_{4,3}$ + $x_{5,3}$ + $x_{6,3}$ + $x_{7,3}$ = 270.0

$x_{0,4}$ + $x_{1,4}$ + $x_{2,4}$ + $x_{3,4}$ + $x_{4,4}$ + $x_{5,4}$ + $x_{6,4}$ + $x_{7,4}$ = 375.0

$x_{0,5}$ + $x_{1,5}$ + $x_{2,5}$ + $x_{3,5}$ + $x_{4,5}$ + $x_{5,5}$ + $x_{6,5}$ + $x_{7,5}$ = 225.0

$x_{0,6}$ + $x_{1,6}$ + $x_{2,6}$ + $x_{3,6}$ + $x_{4,6}$ + $x_{5,6}$ + $x_{6,6}$ + $x_{7,6}$ = 185.0

$x_{0,7}$ + $x_{1,7}$ + $x_{2,7}$ + $x_{3,7}$ + $x_{4,7}$ + $x_{5,7}$ + $x_{6,7}$ + $x_{7,7}$ = 110.0 

Each $x_{i,j} \geq$ 0

In general,

$ minimize \sum_{m=1}^8\sum_{w=1}^8 C(w,m)*x_{wm} $\
subject to constraints 

$\sum_{w=1}^8x_{wm}=d_m \forall m \in \{ 1,2,....8 \}$

$ \sum_{m=1}^8x_{wm} \leq s_w \forall w \in\{1,2,....8\}$

$ x_{wm}\geq 0\ \forall\ w,m \in \{1,2,....,8\}$

where m denotes the number of markets and w denotes number of warehouses. $x_{wm}$ denotes the number of truckloads between mth market and wth warehouse. $d_{m}$ is the demand at mth market and $s_{w}$ is the available stock of wth warehouse. C(w,m) is the transportation cost of sending a truckload of EVs from warehouse w to market m.

**Answer 7:**

Total cost =  2192870.0

Number of truckloads of EVs between Ahmedabad and Jodhpur : 300.0

Number of truckloads of EVs between Ahmedabad and Dibrugarh : 45.0

Number of truckloads of EVs between Ahmedabad and Raipur : 5.0

Number of truckloads of EVs between Bengaluru and Trichy : 120.0

Number of truckloads of EVs between Bengaluru and Vijayawada : 65.0

Number of truckloads of EVs between Chennai and Trichy : 305.0

Number of truckloads of EVs between Delhi and Simla : 95.0

Number of truckloads of EVs between Delhi and Dibrugarh : 85.0

Number of truckloads of EVs between Kolkata and Dibrugarh : 95.0

Number of truckloads of EVs between Kolkata and Agartala : 110.0

Number of truckloads of EVs between Lucknow and Simla : 175.0

Number of truckloads of EVs between Mumbai and Hassan : 205.0

Number of truckloads of EVs between Mumbai and Raipur : 120.0

Number of truckloads of EVs between Hyderabad and Vijayawada : 310.0

**Answer 8:**

Since the objective of the LP is the minimize cost, we can increase the cost of transportation from Lucknow to Simla so much that the number of truck loads between Lucknow and Simla becomes zero. I have set the cost of transportation between Lucknow and Simla to be 10000000000.

**Answer 9:**

Total cost =  2380570.0

Number of truckloads of EVs between Ahmedabad and Jodhpur : 300.0

Number of truckloads of EVs between Ahmedabad and Raipur : 50.0

Number of truckloads of EVs between Bengaluru and Trichy : 120.0

Number of truckloads of EVs between Bengaluru and Vijayawada : 65.0

Number of truckloads of EVs between Chennai and Trichy : 305.0

Number of truckloads of EVs between Delhi and Simla : 180.0

Number of truckloads of EVs between Kolkata and Dibrugarh : 95.0

Number of truckloads of EVs between Kolkata and Agartala : 110.0

Number of truckloads of EVs between Lucknow and Dibrugarh : 130.0

Number of truckloads of EVs between Lucknow and Raipur : 45.0

Number of truckloads of EVs between Mumbai and Hassan : 205.0

Number of truckloads of EVs between Mumbai and Simla : 90.0

Number of truckloads of EVs between Mumbai and Raipur : 30.0

Number of truckloads of EVs between Hyderabad and Vijayawada : 310.0

Number of truckloads of EVs between Hyderabad and Raipur : 60.0

**Observations:**

Now, Lucknow cannot send any truckload to Simla. Previously, Lucknow was sending 175 truckloads to Simla. Simla still has demand of 270 truckloads which needs to be fulfilled so now it gets the remaining supply (175) from other cities like Delhi and Mumbai. Lucknow has available stock of 175, so it can send the stock which was previously being supplied to Simla (175) to other cities like Dibrugarh and Raipur. Because of these changes, some changes also appear in number of truckloads transported between other cities to satisfy all requirements of demands and available stock.

**Answer 10:**

Since the objective of the LP is the minimize cost, we can increase the cost of transportation from Kolkata to Dibrugarh and Delhi to Raipur so much that the number of truck loads between them becomes zero. I have set the cost of transportation in Kolkata-Dibrugarh link and Delhi-Raipur link to be 10000000000 and 10000000000.

**Answer 11:**

Total cost =  2518200.0

Number of truckloads of EVs between Ahmedabad and Jodhpur : 300.0

Number of truckloads of EVs between Ahmedabad and Dibrugarh : 50.0

Number of truckloads of EVs between Bengaluru and Trichy : 120.0

Number of truckloads of EVs between Bengaluru and Vijayawada : 65.0

Number of truckloads of EVs between Chennai and Trichy : 305.0

Number of truckloads of EVs between Delhi and Simla : 180.0

Number of truckloads of EVs between Kolkata and Raipur : 95.0

Number of truckloads of EVs between Kolkata and Agartala : 110.0

Number of truckloads of EVs between Lucknow and Dibrugarh : 175.0

Number of truckloads of EVs between Mumbai and Hassan : 205.0

Number of truckloads of EVs between Mumbai and Simla : 90.0

Number of truckloads of EVs between Mumbai and Raipur : 30.0

Number of truckloads of EVs between Hyderabad and Vijayawada : 310.0

Number of truckloads of EVs between Hyderabad and Raipur : 60.0

**Observations:**

Now, Kolkata cannot send any truckload to Dibrugarh. Previously, Kolkata was sending 95 truckloads to Dibrugarh. Dibrugarh still has demand of 225 truckloads which needs to be fulfilled so now it gets the remaining supply (95) from other cities like Ahmedabad and Lucknow. Kolkata has available stock of 205, so it can send the stock which was previously being supplied to Dibrugarh (95) to other cities like Raipur.

Delhi was previously also not sending any truckload to Raipur so there will be no difference now.

Because of these changes, some changes also appear in number of truckloads transported between other cities to satisfy all requirements of demands and available stock.